superset/assets/javascripts/logger.js

```diff
@@ -0,0 +1,79 @@
+import $ from 'jquery';
+
+export const LOG_ACTIONS_LOAD_DASHBOARD = 'load_dashboard';
+export const LOG_ACTIONS_LOAD_EXPLORE = 'load_explore';
+export const LOAD_EVENT = 'load';
+
+export class ActionLog {
+  constructor({ action, containerId, eventName }) {
+    this.containerId = containerId;
+    this.action = action;
+    this.eventName = eventName;
+    this.startAt = 0;
+    this.duration = 0;
+    this.events = [];
+
+    this.addEvent = this.addEvent.bind(this);
+  }
+
+  setAttribute(name, value) {
+    this[name] = value;
+  }
+
+  addEvent(eventBody) {
+    this.events.push(eventBody);
+  }
+}
+
+const handlers = {};
+
+export default {
+  start(log) {
+    log.setAttribute('startAt', this.getTimestamp());
+    if (!handlers[log.eventName]) {
+      handlers[log.eventName] = [];
+    }
+    handlers[log.eventName].push(log.addEvent);
+  },
+
+  append(eventName, eventBody) {
+    return handlers[eventName].length &&
+      handlers[eventName].forEach((handler) => {
+        handler(eventBody);
+      });
+  },
+
+  end(log) {
+    log.setAttribute('duration', this.getTimestamp() - log.startAt);
+    this.send(log);
+
+    if (handlers[log.eventName].length) {
+      const index = handlers[log.eventName].findIndex(handler => (handler === log.addEvent));
+      handlers[log.eventName].splice(index, 1);
+    }
+  },
+
+  send(log) {
+    const { action, containerId, events, startAt, duration } = log;
+    const url = `/superset/${action}/?${action === LOG_ACTIONS_LOAD_DASHBOARD ? 'dashboard_id' : 'slice_id'}=${containerId}`;
```

Comment:
this breaks if there are more than two types of load actions yeah? could we make it check each explicitly?

Reply:
i changed the way tracking type of load actions. i added `source` and `sourceId` attributes in ActionLog instance.